R/tag_query.R

```diff
@@ -0,0 +1,1601 @@
+    for (parent in parents) {
+      if (!isTagEnv(parent)) next
+      if (ancestorsMap$has(parent)) next
+      # Mark element
+      ancestorsMap$set(parent, TRUE)
+      # Add to final set
+      pushFn(parent)
+      # Add to next iteration
+      nextStack$push(parent)
+    }
+    els <- nextStack$asList()
+  }
+  ancestorsStack$uniqueList()
+}
+# Return a unique list of the closest ancestor elements that match the css selector
+# Should behave very similarly to ancestors
+tagQueryFindClosest <- function(els, cssSelector = NULL) {
+  if (is.null(cssSelector)) {
+    return(
+      tagQueryFindParent(els, NULL)
+    )
+  }
+  selector <- cssSelectorToSelector(cssSelector)
+  # use the map for `has()` and stack for `values()`
+  ancestorsMap <- envirMap()
+  closestStack <- envirStackUnique()
+
+  # For every element
+  tagQueryWalk(els, function(el) {
+    # Make sure it is a tag environment
+    if (!isTagEnv(el)) return()
+
+    # While traversing up the parents...
+    while (!is.null(el)) {
+      # If the element has been seen before...
+      if (ancestorsMap$has(el)) {
+        # Stop traversing, as any matching parent found would be removed (unique info only)
+        return()
+      }
+      # Mark the ancestor as visited
+      ancestorsMap$set(el, TRUE)
+      # If it is a match...
+      if (elMatchesSelector(el, selector)) {
+        # Add to return value
+        closestStack$push(el)
+        return()
+      }
+      # set to parent element and repeat
+      el <- el$parent
+    }
+  })
+
+  closestStack$uniqueList()
+}
+# Get all unique children tag envs
+tagQueryFindChildren <- function(els, cssSelector = NULL) {
+  childrenStack <- envirStackUnique()
+  pushFn <- pushFnWrapper(childrenStack, cssSelector)
+  tagQueryWalk(els, function(el) {
+    if (!isTagEnv(el)) return()
+    tagQueryWalk(el$children, pushFn)
+  })
+  childrenStack$uniqueList()
+}
+
+# Return all unique siblings of each el in els
+tagQueryFindSiblings <- function(els, cssSelector = NULL) {
+  siblingStack <- envirStackUnique()
+  pushFn <- pushFnWrapper(siblingStack, cssSelector)
+  tagQueryWalk(els, function(el) {
+    if (!isTagEnv(el)) return()
+    elKey <- el$envKey
+    tagQueryWalk(el$parent$children, function(sibling) {
+      if (!isTagEnv(sibling)) return()
+      siblingKey <- sibling$envKey
+      if (elKey != siblingKey) {
+        pushFn(sibling)
+      }
+    })
+  })
+  siblingStack$uniqueList()
+}
+
+# Filter the selected elements using a function
+# The answer of `fn(el, i)` should work in an `if` block
+tagQueryFindFilter <- function(els, fn) {
+  if (is.character(fn)) {
+    selector <- cssSelectorToSelector(fn)
+    fn <- function(el, i) {
+      elMatchesSelector(el, selector)
+    }
+  }
+  validateFnCanIterate(fn)
+
+  filterStack <- envirStack()
+  selectedWalkI(els, function(el, i) {
+    if (fn(el, i)) {
+      filterStack$push(el)
+    }
+  })
+
+  filterStack$asList()
+}
+
+
+
+cssSelectorToSelector <- function(cssSelector) {
```

Comment:
It would help to have a short explanation of what this function is supposed to do.